lib/discrete_maths/logic_operations.py

```diff
@@ -0,0 +1,178 @@
+    if not proposition:
+        raise SyntaxError("Empty proposition")
+
+    for char in proposition:
+        if char not in ['~', '(', ')', 'v', '^', '>', '=', ' ']:
+            if not char.isalpha():
+                raise ValueError("Invalid character in proposition: " + char)
+    
+    if not is_balanced_proposition(proposition):
+        raise SyntaxError("Unbalanced proposition")
+    
+    if not is_well_formed_proposition(proposition):
+        raise SyntaxError("Proposition is not well-formed")
+    
+    return True
+    
+# Calculate the result of a proposition given a set of variable values and an RPN version of the proposition
+def calculate_rpn_prop(rpn: list, variable_values: dict) -> bool:
+    stack = []
+
+    for token in rpn:
+        if token in variable_values.keys():
+            stack.append(variable_values[token])
+        elif token == "~":
+            stack.append(negation(stack.pop()))
+        elif token == "^":
+            stack.append(conjunction(stack.pop(), stack.pop()))
+        elif token == "v":
+            stack.append(disjunction(stack.pop(), stack.pop()))
+        elif token == ">":
+            stack.append(implies(stack.pop(), stack.pop()))
+        elif token == "=":
+            stack.append(iff(stack.pop(), stack.pop()))
+    
+    return stack.pop()
+
+# Create a list of lists of all the possible values for the variables in the rpn proposition
+def create_variable_values(rpn: list) -> list:
+    variable_list = []
+    variable_count = 0
+    for token in rpn:
+        if token.isalpha() and token not in variable_list and token not in OP_PRECIDENCE.keys():
+            variable_list.append(token)
+            variable_count += 1
+
+    combinations_list = []
+    for i in range(2 ** variable_count):
+        combination = []
+        for j in range(variable_count):
+            combination.append(bool(i & (1 << j)))
+        combinations_list.append(combination)
+    
+    return combinations_list, variable_list
+
+# Check if a proposition is a tautology
+def is_prop_tautolgy(proposition: str) -> bool:
+    # Sanity check
+    if not is_valid_proposition(proposition):
+        raise SyntaxError("Invalid proposition")
+
+    proposition = proposition.replace(" ", "")
+    rpn_tokenization = shunting_yard(proposition, OP_PRECIDENCE)
+
+    combinations_list, variable_list = create_variable_values(rpn_tokenization)
+
+    # Calculate the result of the proposition for each combination and add the result to the list
+    results = []
+    for combination in combinations_list:
+        variable_values = {}
+        for i in range(len(variable_list)):
+            variable_values[variable_list[i]] = combination[i]
+        results.append(calculate_rpn_prop(rpn_tokenization, variable_values))
+
+    # Check if all the results are true
+    return all(results)
+
+# Create a truth table from a proposition
+def create_truth_table(proposition: str) -> str:
```

Comment:
The logic that creates the table shouldn't be here. It should be in the cog. However, the logic that evaluates the logical proposition can/should stay here. 

The reason why is that this library shouldn't care too much about how the evaluation of a proposition is used. It should only care about evaluating it and returning some results. Who cares about how it's used and what is displayed is the cog. The cog is the link between the logic and the user. 

In other words, all the sanity checkers and logic for the evaluation is fine here but the code to draw (this includes making the variables) the table shouldn't be here but in the cog.